### PR TITLE
Feature/1 Initial implementation

### LINKED ISSUE #1: Initial Implementation
Create an initial Implementation

### PR DESCRIPTION
**Pull request by jakobbraun**
Closes #1

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+[![Build Status](https://github.com/exasol/java-class-list-extractor/actions/workflows/ci-build.yml/badge.svg)](https://github.com/exasol/java-class-list-extractor/actions/workflows/ci-build.yml)
+
+The Java Class List Extractor is a tool that helps you to capture a list of loaded classed from Java UDFs / AdapterScripts in an Exasol database. The class list is required for improving the start-up performance of Java UDF using [java-udf-startup-time-improver](https://github.com/exasol/java-udf-startup-time-improver/).
+
+This tool consists of two parts:
+
+* The `ClassListExtractor`
+* The `ClassListVerifier`
+
+The extractor captures the class list of a UDF run. The verifier checks that this class list is embedded into the JAR archive of the application. If it's not it breaks the build and writes the expected class list to `target/expeced-classes.lst`.From there you can copy it to your resources and restart the build.
```
Comment (kaklakariada) [inline README.md]:
```suggestion
The extractor captures the class list of a UDF run. The verifier checks that this class list is embedded into the JAR archive of the application. If it's not it breaks the build and writes the expected class list to `target/expeced-classes.lst`. From there you can copy it to your resources and restart the build.
```

File: README.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+## What to Optimize for
+
+The class list determines which classes will be preloaded. That cna reduce the start-up time og the JVM. However, if you preload many classes that are not used at runtime that's an overhead. So you should make a conscious decision which test cases you want to analyze. Analyzing all test-cases (including lots of tests for unlikely edge cases) gives a non-ideal result.
```
Comment (kaklakariada) [inline README.md]:
```suggestion
The class list determines which classes will be preloaded. This can reduce the start-up time of the JVM. However, if you preload many classes that are not used at runtime that's an overhead. So you should make a conscious decision which test cases you want to analyze. Analyzing all test-cases (including lots of tests for unlikely edge cases) gives a non-ideal result.
```

File: README.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+## Usage
+
+* Add this project as maven test dependency.
+* Create a ClassListExtractor:
+    ```
+    final ClassListExtractor extractor = new ClassListExtractor(EXASOL.getDefaultBucket(), port -> new InetSocketAddress(EXASOL.getHostIp() + "", port));
+    ```
+* Make sure that you close the extractor. For example in the `afterAll` if you created it in the `beforeAll()`:
```
Comment (kaklakariada) [inline README.md]:
```suggestion
* Make sure that you close the extractor. For example in `afterAll()` if you created it in `beforeAll()`:
```

File: README.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    connection.close();
+    extractor.close();
+  }
+  ```
+* Add the JVM options provided by `extractor.getJvmOptions()` to your UDF or Adapter Script definitions.
+
+  If you use the [test-db-builder-java](https://github.com/exasol/test-db-builder-java), you can use the following snippet:
+  ```
+  final ExasolObjectFactory objectFactory = new ExasolObjectFactory(connection,
+     ExasolObjectConfiguration.builder().withJvmOptions(jextractor.getJvmOptions()).build());
```
Comment (kaklakariada) [inline README.md]:
```suggestion
  final ExasolObjectFactory objectFactory = new ExasolObjectFactory(connection,
     ExasolObjectConfiguration.builder().withJvmOptions(extractor.getJvmOptions()).build());
```

File: README.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  ```
+  If you also use [udf-debugging-java](https://github.com/exasol/udf-debugging-java/), use:
+  ```
+  final List<String> jvmOptions = new ArrayList<>(Arrays.asList(udfTestSetup.getJvmOptions()));
+      jvmOptions.addAll(Arrays.asList(extractor.getJvmOptions()));
+      final ExasolObjectFactory objectFactory = new ExasolObjectFactory(connection,
+         ExasolObjectConfiguration.builder().withJvmOptions(jvmOptions.toArray(String[]::new)).build());
```
Comment (kaklakariada) [inline README.md]:
```suggestion
  final List<String> jvmOptions = new ArrayList<>(Arrays.asList(udfTestSetup.getJvmOptions()));
  jvmOptions.addAll(Arrays.asList(extractor.getJvmOptions()));
  final ExasolObjectFactory objectFactory = new ExasolObjectFactory(connection,
         ExasolObjectConfiguration.builder().withJvmOptions(jvmOptions.toArray(String[]::new)).build());
```

File: README.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+     try (final ResultSet result = statement.executeQuery("SELECT TEST.MY_UDF() AS RES")) {
+       result.next();
+       assertThat(result.getString("RES"), equalTo("Hello Java World!"));
+     }
+  });
+  ```
+* Add the verifier:
+  ```
+  new ClassListVerifier()
+    .verifyClassListFile(classList, Path.of("target(my-app.jar"));
```
Comment (kaklakariada) [inline README.md]:
```suggestion
  new ClassListVerifier()
    .verifyClassListFile(classList, Path.of("target/my-app.jar"));
```

File: README.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  ```
+
+If you've multiple queries that you want to optimize for, you can either run all of them in the same capture or run capture multiple times. In the second case you have to take care for merging the generated class lists. We recommend using a `Set` for that task. That makes sure that there are no duplicate entries.
```
Comment (kaklakariada) [inline README.md]:
```suggestion
If you've multiple queries that you want to optimize for, you can either run all of them in the same `capture()` or run `capture()` multiple times. In the second case you have to take care for merging the generated class lists. We recommend using a `Set` for that task. That makes sure that there are no duplicate entries.
```

File: agent/src/main/java/com/exasol/classlistextractor/agent/ClassListExtractorAgent.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+import java.nio.channels.SocketChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+import com.exasol.errorreporting.ExaError;
+
+/**
+ * This class is a Java agent that can be attached to the JVM in a UDF. It installs a hook on shutdown of the JVM and
+ * then transfers the contents of /tmp/classes.lst to a tcp socket. This agent is required since we can't access to the
+ * /tmp of a UDF.
```
Comment (kaklakariada) [inline agent/src/main/java/com/exasol/classlistextractor/agent/ClassListExtractorAgent.java]:
```suggestion
 * then transfers the contents of /tmp/classes.lst to a tcp socket. This agent is required since we can't access the
 * /tmp of a UDF.
```

File: agent/src/main/java/com/exasol/classlistextractor/agent/ClassListExtractorAgent.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+     * @param argument argument string
+     */
+    public static void premain(final String argument) {
+        final InetSocketAddress serverAddress = parseArgumentString(argument);
+        Runtime.getRuntime().addShutdownHook(new Thread(new ClassListSender(serverAddress)));
+    }
+
+    private static InetSocketAddress parseArgumentString(final String argument) {
+        if (argument == null) {
+            throw getInvalidArgsException("<missing>");
```
Comment (kaklakariada) [inline agent/src/main/java/com/exasol/classlistextractor/agent/ClassListExtractorAgent.java:33]: Maybe add an error message built with `ExaError`.

File: agent/src/test/java/com/exasol/classlistextractor/agent/ClassListExtractorAgentIT.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    @Test
+    void testGetClassList(@TempDir final Path tempDir) throws IOException, InterruptedException {
+        runCommand(new String[] { "javac", Path.of(HELLO_WORLD_CLASS).toString(), "-d", tempDir.toString() });
+        final SimpleServer simpleServer = new SimpleServer();
+        new Thread(simpleServer).start();
+        final Path coverage = Path.of("target/jacoco-it-nested.exec").toAbsolutePath();
+        runCommand(new String[] { "java", "-XX:DumpLoadedClassList=/tmp/classes.lst",
+                "-javaagent:target/jacoco-agent/org.jacoco.agent-runtime.jar=destfile=" + coverage,
+                "-javaagent:" + AGENT_JAR + "=localhost:" + simpleServer.serverSocket.getLocalPort(), "-cp",
+                tempDir.toString(), "com/exasol/classlistextractor/agent/HelloWorld" });
```
Comment (kaklakariada) [inline agent/src/test/java/com/exasol/classlistextractor/agent/ClassListExtractorAgentIT.java]: Why do you need Jacoco here?

File: verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListExtractor.java [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+import java.net.InetSocketAddress;
+import java.util.List;
+import java.util.concurrent.TimeoutException;
+
+import com.exasol.bucketfs.Bucket;
+import com.exasol.bucketfs.BucketAccessException;
+import com.exasol.errorreporting.ExaError;
+
+/**
+ * This class captures the classes loaded the JVM of a UDF.
```
Comment (kaklakariada) [inline verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListExtractor.java]:
```suggestion
 * This class captures the classes loaded by the JVM of a UDF.
```

File: verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListExtractor.java [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+     */
+    public ClassListExtractor(final Bucket bucket, final LocalServiceExposer serviceExposer) {
+        uploadAgent(bucket);
+        this.server = new ClassListServer();
+        this.inDbServerAddress = serviceExposer.exposeLocalServiceToDatabase(this.server.getPort());
+        this.agentInBucket = "/buckets/" + bucket.getBucketFsName() + "/" + bucket.getBucketName() + "/" + AGENT_JAR;
+    }
+
+    /**
+     * Get the JVM options for that must be added to the script.
```
Comment (kaklakariada) [inline verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListExtractor.java]:
```suggestion
     * Get the JVM options that must be added to the script.
```

File: verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListExtractor.java [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+
+    private void assertGetJvmOptionsWasCalled() {
+        if (!this.getJvmOptionsWasCalled) {
+            throw new IllegalStateException(ExaError.messageBuilder("E-JCLE-VF-11")
+                    .message("Missing call to getJvmOptions().")
+                    .mitigation(
+                            "Make sure that you add the JVM options provided by getJvmOptions to your script definition.")
+                    .toString());
+        }
+    }
```
Comment (kaklakariada) [inline verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListExtractor.java:92]: Very smart 👍

File: verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListExtractor.java [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+
+    private InputStream getAgentJarAsStream() {
+        final InputStream stream = getClass().getClassLoader().getResourceAsStream(AGENT_JAR);
+        if (stream == null) {
+            throw new IllegalStateException(ExaError.messageBuilder("F-JCLE-VF-1")
+                    .message("Could not access agent jar " + AGENT_JAR + ".")
+                    .mitigation("The jar should be bundled with this package. "
+                            + "If this happens in production please report an issue. "
+                            + "If it happens when testing the verifier itself make sure you ran mvn package before.")
```
Comment (kaklakariada) [inline verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListExtractor.java:110]: 👍

File: verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListExtractor.java [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+        return stream;
+    }
+
+    @Override
+    public void close() {
+        this.server.close();
+    }
+
+    /**
+     * Interface for a method that runs a query to a script, that's class loadig you want to analyze with
```
Comment (kaklakariada) [inline verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListExtractor.java]:
```suggestion
     * Interface for a method that runs a query to a script, that loads classes you want to analyze with
```

File: verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListExtractor.java [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+     * {@link #capture(CallbackWithQueryToScript)}.
+     */
+    @FunctionalInterface
+    public interface CallbackWithQueryToScript {
+        /**
+         * Method that runs a query to a script, that's class loadig you want to analyze with
```
Comment (kaklakariada) [inline verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListExtractor.java]:
```suggestion
         * Method that runs a query to a script, that loads classes you want to analyze with
```

File: verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListServer.java [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+     */
+    public void clear() {
+        this.clientHandler.clear();
+    }
+
+    /**
+     * Get the received class list.
+     * <p>
+     * The class list is a union (with duplicate removal) of all revieved class lists since the last call to
+     * {@link #clear()}.
```
Comment (kaklakariada) [inline verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListServer.java]:
```suggestion
     * The class list is a union (with duplicates removed) of all retrieved class lists since the last call to
     * {@link #clear()}.
```

File: verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListServer.java [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+            try {
+                this.serverSocket.close();
+            } catch (final IOException exception) {
+                throw new UncheckedIOException(ExaError.messageBuilder("E-JCLE-VF-6")
+                        .message("Failed to close the server for retrieving the class lists of a project.").toString(),
+                        exception);
+            }
+        }
+
+        synchronized void clear() {
```
Comment (kaklakariada) [inline verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListServer.java]: This is the only `synchronized` method. Did you mean to use a `synchronized (this.serverThreadMonitor) {}` block like in the other methods?

File: verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListVerifier.java [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+import java.nio.file.Path;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+
+import com.exasol.errorreporting.ExaError;
+
+/**
+ * This class verifies that a jar file contains an {@code classes.lst} file and validates that it contains the expected
```
Comment (kaklakariada) [inline verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListVerifier.java]:
```suggestion
 * This class verifies that a jar file contains a {@code classes.lst} file and validates that it contains the expected
```

File: verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListVerifier.java [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+ * classes.
+ */
+public class ClassListVerifier {
+    private static final String CLASSES_LIST_FILE_NAME = "classes.lst";
+
+    /**
+     * Verify that a jar file contains an {@code classes.lst} file and validates that it contains the expected classes.
```
Comment (kaklakariada) [inline verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListVerifier.java]:
```suggestion
     * Verify that a jar file contains a {@code classes.lst} file and validates that it contains the expected classes.
```

File: verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListVerifier.java [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+                if (CLASSES_LIST_FILE_NAME.equals(entry.getName())) {
+                    return Optional.of(new String(zipInputStream.readAllBytes(), StandardCharsets.UTF_8));
+                }
+                entry = zipInputStream.getNextEntry();
+            }
+            return Optional.empty();
+        } catch (final IOException exception) {
+            throw new UncheckedIOException(ExaError.messageBuilder("E-JCLE-VF-12")
+                    .message("Error while reading jar-file contents for checking if it contains the correct "
+                            + CLASSES_LIST_FILE_NAME + " file.")
```
Comment (kaklakariada) [inline verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListVerifier.java]:
```suggestion
                    .message("Error while reading jar-file {{jar file}} contents for checking if it contains the correct "
                            + CLASSES_LIST_FILE_NAME + " file.", jarFile)
```

File: verifier/src/main/java/com/exasol/classlistextractor/verifier/LocalServiceExposer.java [hunk 22]
```diff
@@ -0,0 +1,18 @@
+package com.exasol.classlistextractor.verifier;
+
+import java.net.InetSocketAddress;
+
+/**
+ * Implementors of this interface exposes a local service (socket) into the Exasol database.
```
Comment (kaklakariada) [inline verifier/src/main/java/com/exasol/classlistextractor/verifier/LocalServiceExposer.java]:
```suggestion
 * Implementors of this interface expose a local service (socket) into the Exasol database.
```

File: verifier/src/test/java/com/exasol/classlistextractor/verifier/ClassListExtractorIT.java [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+    private static Statement statement;
+    private static ClassListExtractor extractor;
+
+    @BeforeAll
+    static void beforeAll() throws SQLException {
+        connection = EXASOL.createConnection();
+        statement = connection.createStatement();
+        final UdfTestSetup udfTestSetup = new UdfTestSetup(EXASOL.getHostIp(), EXASOL.getDefaultBucket(), connection);
+        extractor = new ClassListExtractor(EXASOL.getDefaultBucket(),
+                port -> new InetSocketAddress(EXASOL.getHostIp() + "", port));
```
Comment (kaklakariada) [inline verifier/src/test/java/com/exasol/classlistextractor/verifier/ClassListExtractorIT.java]:
```suggestion
                port -> new InetSocketAddress(EXASOL.getHostIp(), port));
```
❓

**Review by kaklakariada — CHANGES_REQUESTED** [review]
Readable & clean code and very good test coverage 👍

File: agent/src/test/java/com/exasol/classlistextractor/agent/ClassListExtractorAgentIT.java (hunk 10 above)
Comment (jakobbraun) [reply agent/src/test/java/com/exasol/classlistextractor/agent/ClassListExtractorAgentIT.java]: for getting the test coverage. Since it runs in another JVM jacoco would not detect it.

File: agent/src/main/java/com/exasol/classlistextractor/agent/ClassListExtractorAgent.java (hunk 9 above)
Comment (jakobbraun) [reply agent/src/main/java/com/exasol/classlistextractor/agent/ClassListExtractorAgent.java:33]: it's built in getInvalidArgsException.

Comment (kaklakariada) [reply agent/src/main/java/com/exasol/classlistextractor/agent/ClassListExtractorAgent.java:33]: 👍

File: agent/src/test/java/com/exasol/classlistextractor/agent/ClassListExtractorAgentIT.java (hunk 10 above)
Comment (kaklakariada) [reply agent/src/test/java/com/exasol/classlistextractor/agent/ClassListExtractorAgentIT.java]: 👍

File: verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListServer.java (hunk 18 above)
Comment (jakobbraun) [reply verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListServer.java]: good catch. It's was outdated.

File: agent/src/test/java/com/exasol/classlistextractor/agent/ClassListSenderTest.java [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+class ClassListSenderTest {
+
+    @TempDir
+    Path tempDir;
+
+    @Test
+    @SuppressWarnings("java:S2925") // sleep is a pragmatic solution here
```
Comment (kaklakariada) [inline agent/src/test/java/com/exasol/classlistextractor/agent/ClassListSenderTest.java:24]: 👍

**Review by kaklakariada — DISMISSED** [review]

**Review by kaklakariada — APPROVED** [review]
